application/src/test/java/org/thingsboard/server/controller/BaseDashboardControllerTest.java

```diff
@@ -495,4 +514,107 @@ public void testAssignDashboardToEdge() throws Exception {
         Assert.assertEquals(0, pageData.getData().size());
     }
 
+    @Test
+    public void testDeleteDashboardExistsRelation_Error_RestoreRelation_DeleteRelation_DeleteDashboard_Ok() throws Exception {
+        Dashboard dashboard = new Dashboard();
+        dashboard.setTitle("My dashboard");
+        Dashboard savedDashboard = doPost("/api/dashboard", dashboard, Dashboard.class);
+
+        Alarm savedAlarm = createAlarm(savedDashboard.getId());
+
+        AlarmOperationResult alarmOperationResult = alarmService.createOrUpdateAlarm(savedAlarm);
+        Assert.assertEquals("List of propagatedEntities is not equal to number of created propagatedEntities!",
+                1, alarmOperationResult.getPropagatedEntitiesList().size());
+        Assert.assertEquals("DashboardId in propagatedEntities is not equal savedDashboardId!",
+                savedDashboard.getId(), alarmOperationResult.getPropagatedEntitiesList().get(0));
+
+        DeviceProfile deviceProfile =  createDeviceProfile();
+        deviceProfile.setDefaultDashboardId(savedDashboard.getId());
+        DeviceProfile savedDeviceProfile = doPost("/api/deviceProfile", createDeviceProfile(), DeviceProfile.class);
+
+        savedDeviceProfile.setDefaultDashboardId(savedDashboard.getId());
+        DeviceProfile updatedDeviceProfile = doPost("/api/deviceProfile", savedDeviceProfile, DeviceProfile.class);
+
+        String typeRelation = EntityRelation.CONTAINS_TYPE;
+        EntityRelation relation = new EntityRelation(updatedDeviceProfile.getId(), savedDashboard.getId(), typeRelation);
+        doPost("/api/relation", relation).andExpect(status().isOk());
+        String url = String.format("/api/relation?fromId=%s&fromType=%s&relationType=%s&toId=%s&toType=%s",
+                updatedDeviceProfile.getUuidId(), EntityType.DEVICE_PROFILE,
+                typeRelation, savedDashboard.getUuidId(), EntityType.DASHBOARD
+        );
+        EntityRelation foundRelation = doGet(url, EntityRelation.class);
+        Assert.assertNotNull("Relation is not found!", foundRelation);
+        Assert.assertEquals("Found relation is not equals origin!", relation, foundRelation);
+
+        String dashboardIdStr = savedDashboard.getId().getId().toString();
+        doDelete("/api/dashboard/" + dashboardIdStr)
+                .andExpect(status().isBadRequest())
+                .andExpect(statusReason(containsString("The dashboard referenced by the device profiles cannot be deleted!")));
+
+        Dashboard afterErrorDeleteDashboard = doGet("/api/dashboard/" + dashboardIdStr,  Dashboard.class);
+        Assert.assertNotNull("Dashboard is not found!", afterErrorDeleteDashboard);
+        Assert.assertEquals("Dashboard after delete error is not equals origin!", savedDashboard, afterErrorDeleteDashboard);
+
+        EntityRelation afterErrorDeleteDashboardRelation = doGet(url, EntityRelation.class);
+        Assert.assertNotNull("Relation is not found!", foundRelation);
```

Comment:
It seems that instead `foundRelation` you want to assert the `afterErrorDeleteDashboardRelation`

Reply:
Message `Relation is not found` on assertion is expected to be unique. It might be confusing to detect the problem by non-unique message. Please, add some sensible suffix to the message like 'Relation not found after dashboard deletion bad request'

Reply:
Ok